### ISSUE TITLE
toIterable() fails with completed iterators

### ISSUE THREAD

**Issue by mattbishop** [vitaly-t/iter-ops#204]
I have a case where I need to wrap a possibly-completed iterator as an iterator to pipe() through some other things. I know the result will be an empty iterator, but I don't have a way to test for this before building the pipe() chain.

In debugging, I found that toIterable() calls a typeguard function isIteratorResult: https://github.com/vitaly-t/iter-ops/blob/35bc52e6844ad217ea30a95d1a0c48be479defb0/src/typeguards.ts#L91 and fails to see the next() result as an iterator result. It wraps it as a SingleIterator, which breaks my app because it isn't expecting an iterator result.

`isIteratorResult` tests for `value`, but a completed iterator does not return a value. It does return `done` in all cases:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Iterators_and_Generators#iterators

The fix is to look for a `done` property instead of a `value` property.

**Comment by vitaly-t** [comment]
It may be argued that `isIteratorResult` should return `return has(value, 'value') || has(value, 'done')`, however the bug is triggered by `has` implementation, which returns false for `value: undefined`, which is definitely incorrect.

The proper syntax goes like this: `value` must be always present, and `done` is optional. However, all JavaScript implementations feature both at all times, i.e. the last element is always `{value: undefined, done: true}`.

Let's look at how `IteratorResult` is defined within ES2015 that we are using:

```ts
interface IteratorYieldResult<TYield> {
    done?: false;
    value: TYield;
}

interface IteratorReturnResult<TReturn> {
    done: true;
    value: TReturn;
}

type IteratorResult<T, TReturn = any> = IteratorYieldResult<T> | IteratorReturnResult<TReturn>;
```

So we can see that `done` is optional while inside generators, which is odd, but ok. That's why I think that fixing `has` implementation is more appropriate.

@RebeccaStevens What do you think? If you agree, please make the change :wink:

**Comment by vitaly-t** [comment]
@RebeccaStevens b.t.w., you have this comment inside `has` - `// Cannot use `in` operator as it doesn't work for strings.`

Perhaps [hasOwnProperty](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty) is usable for this? ;)

**Comment by RebeccaStevens** [comment]
`Object.prototype.hasOwnProperty` (or `Object.hasOwn`) is usually my go to for this, but it doesn't work in this case.
In general, we need to know if the object has the key, regardless of if it's direct or inherited. Not for the case of `isIteratorResult`, but for the other functions such as `isSyncIterable` and `isPromiseLike`.
Additionally, that function doesn't work for symbols (`Object.getOwnPropertySymbols` would need to be used for symbols, but again, that's for direct symbols).

I deicide it is best to go back to using the `in` operator. If we convert the object to an `Object` first, then it should work as desired.

**Comment by vitaly-t** [comment]
[v2.8.0-beta.2](https://github.com/vitaly-t/iter-ops/releases/tag/2.8.0-beta.2) now contains the fix.

I'm gonna play a little with the new [concurrencyFork](https://vitaly-t.github.io/iter-ops/functions/concurrencyFork) operator within [iter-ops-extras](https://github.com/vitaly-t/iter-ops-extras), to make sure it is as every bit as good as I was expecting, and then will make a proper 2.8.0 release :wink:

@RebeccaStevens Never got your view on the new `concurrencyFork` addition. I think it is a game changer. What do you reckon?

**Comment by RebeccaStevens** [comment]
It looks good. I'll have a more in-depth look at it a bit later.

**Comment by vitaly-t** [comment]
The fix for this was released in [v2.8.0](https://github.com/vitaly-t/iter-ops/releases/tag/2.8.0).

@RebeccaStevens I tested it out (`concurrencyFork`) thoroughly, and it was good, so I have released it ;)

**Comment by mattbishop** [comment]
Verified fix in 2.8.0. Thanks!
